pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,298 @@
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchema defines the schema of a group of flows. Note that a flow makes up of a set of inbound API requests with
+// similar attributes and is identified by a pair of strings: the name of the FlowSchema and a "flow distinguisher".
+type FlowSchema struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec
+	// Current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// MatchingPriority controls the ordering the flow-schemas in the cluster. Its value must be greater than zero and
+	// also unique. The primary sort is based on MatchingPriority, ordered highest to lowest (20 before 10).
+	MatchingPriority int32
+	// RequestPriorityRef can reference a RequestPriority in the cluster. If the reference cannot be resolved, the
+	// flow-schema will ignored and marked as invalid in its status.
+	RequestPriorityRef RequestPriorityRef
+	// Distinguisher defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
+	// +optional
+	Distinguisher *FlowDistinguisher
+	// Rules holds all the RuleWithUser for this FlowSchema.
+	Rules []RuleWithUser
+}
+
+// FlowDistinguisher specifies the configuration of the distinguisher.
+type FlowDistinguisher struct {
+	// Source defines the source attribute we adopt to compute the identifier for the flow.
+	Source FlowDistinguisherSource
+}
+
+// FlowDistinguisherSource defines the input source of distinguishing flows.
+type FlowDistinguisherSource string
+
+// RequestPriorityRef contains information that points to the request-priority being used.
+type RequestPriorityRef struct {
+	// APIGroup is the group for the resource being referenced.
+	APIGroup string
+	// Kind is the type of resource being referenced.
+	Kind string
```

Comment:
Kind and Resource are not interchangeable. Notice that CRDs use resources to avoid having to a restmapping, but ownerRefs require a RESTMapping conversion which is lossy and potentially conflicting.  I think we should avoid the need for a RESTMapping.

Reply:
removed `APIGroup` and  `Kind` from the ref object to avoid RESTMapping. now we reference merely by name